app/models/katello/upstream_consumer.rb

```diff
@@ -0,0 +1,61 @@
+require 'katello/resources/candlepin'
+
+module Katello
+  class UpstreamConsumer < OpenStruct
+    CP_CONSUMER = Resources::Candlepin::UpstreamConsumer
+
+    class << self
+      def bind_entitlements(organization, params)
+        CP_CONSUMER.organization = organization
+        params.fetch(:pools).map do |pool|
+          JSON.parse(CP_CONSUMER['entitlements'].post(nil, params: map_pool_attributes(pool)))
```

Comment:
You are using map_pool_attributes to filter out nil pools. Consider doing something like this to clean up the input a bit

```
params[:pools].compact.uniq.map do |pool|
```

Reply:
This JSON parsing seems like it should be in the resource::candlepin layer, not the model layer.

Reply:
I am using `map_pool_attributes` to map our API to candlepin's, as the name suggests, and thought a nil check would be prudent.

I would argue that (while it doesn't matter that much), the way I've written it is more efficient because it loops over the pools once instead of twice (O(N) vs. O(2N)). In addition, I'd argue the explicit nil check is more readable, especially since it happens in a separate method rather than in the same method chain.

Furthermore! ;) We want users to be able to specify multiple of the same pool, so users can get N entitlements of the same pool in one request (this is that thing we were discussing where users sometimes want N entitlements of quantity 1 rather than 1 entitlement of quantity N), so we wouldn't want to `uniq` our pools array.

Reply:
I'm following you on the multiple pools per request so ignire my .uniq, but the current impl could still send nils to the backend:

map_pool_attributes(nil) => nil